pkg/api/steve/collections/collections.go

```diff
@@ -0,0 +1,699 @@
+	logrus.Tracef("[%s] handling request %s\n", apiServiceName, r.URL.Path)
+	vars := mux.Vars(r)
+	op := vars["op"]
+	var projectsOrNamespaces []string
+	if _, ok := vars[projectsOrNamespacesVar]; ok {
+		projectsOrNamespaces = strings.Split(vars[projectsOrNamespacesVar], ",")
+	}
+	namespaces := make(map[string]*corev1.Namespace)
+	resource, err := h.gvrFromVars(vars)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	resourceClient := h.dynamicClient.Resource(resource)
+	if len(projectsOrNamespaces) > 0 {
+		for _, n := range projectsOrNamespaces {
+			ns, _ := h.namespaceCache.Get(n)
+			namespaces[n] = ns
+		}
+	}
+	// get namespaces that don't have the projectID label
+	selector, _ := labels.Parse("!" + nslabels.ProjectIDFieldLabel)
+	nss, err := h.namespaceCache.List(selector)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	finalNamespaces := make(map[string]*corev1.Namespace)
+	for _, ns := range nss {
+		// no namespaces in the selector so all can be included
+		if len(namespaces) == 0 {
+			finalNamespaces[ns.Name] = ns
+			continue
+		}
+		// only include resources in namespaces specified in the selector
+		if op != "!" {
+			if n, ok := namespaces[ns.Name]; ok && n != nil {
+				finalNamespaces[ns.Name] = ns
+			}
+		} else { // only include resources in namespaces NOT specified in the selector
+			if _, ok := namespaces[ns.Name]; !ok {
+				finalNamespaces[ns.Name] = ns
+			}
+		}
+	}
+	namespaces = finalNamespaces
+
+	resp, err := h.getResourcesForNamespaces(r, resource, resourceClient, namespaces)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	returnResp(w, resp)
+}
+
+// scopedHandler handles requests for resources in namespaces in one particular project.
+// steve sends requests to individual namespace endpoints based on the user's rolebindings,
+// so this endpoint has to handle both projects and namespaces in the selector even though
+// only namespaces in the one project will be returned.
+func (h *handler) scopedHandler(w http.ResponseWriter, r *http.Request) {
+	logrus.Tracef("[%s] handling request %s\n", apiServiceName, r.URL.Path)
+	vars := mux.Vars(r)
+	project := vars["project"]
+	op := vars["op"]
+	var projects []string
+	var projectsOrNamespaces []string
+	if _, ok := vars[projectsOrNamespacesVar]; ok {
+		projectsOrNamespaces = strings.Split(vars[projectsOrNamespacesVar], ",")
+	}
+	namespaces := make(map[string]*corev1.Namespace)
+	resource, err := h.gvrFromVars(vars)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	resourceClient := h.dynamicClient.Resource(resource)
+	if len(projectsOrNamespaces) > 0 {
+		projects, namespaces, err = h.projectsAndNamespaces(projectsOrNamespaces)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+	if op != "!" {
```

Comment:
(This comment applies for the general filtering logic in the if/else block below).

I think this filtering logic is a bit complex and has enough edge cases that it will be tough to maintain over time.

What I would recommend instead is logic that does the following:
- First, get the project and all associated namespaces (you can use an indexer to map namespace to owning project so that this lookup is ok from a speed perspective)
- Then If you've requested a filter (through `pjorns`), apply as below:
  - If we are using a positive filter (i.e. =), only include namespaces from the project that meet the filter (i.e. only include stuff from the first step if it matches at least one entry)
  - If we are using a negative filter (i.e. !=) remove those namespaces from the project that meet the filter (i.e. remove stuff from the first step if it matches at least one entry)
  - Don't do any filtering on project (this is controversial, but I think that if you are accessing this endpoint for a project we should return stuff for the project and only for that project - in this context, filtering by project has no meaning). What this means in practice is that if we encounter a value passed in by `pjorns` that doesn't match one of our active namespaces, we ignore that parameter. 
  
I'm hoping that this will make things a bit more simple, but it's possible I'm missing a desired behavior, so please LMK what you think.

Reply:
> I think this filtering logic is a bit complex and has enough edge cases

Yes, the edge cases are why it is complex

> First, get the project and all associated namespaces

This is done on line 378 https://github.com/rancher/rancher/pull/40155/files#diff-ff7986e4e885b7d6b19f5c52f21f48e75b503e7acc41534744ddbc6b5bf68cbcR378

> If we are using a positive filter (i.e. =), only include namespaces from the project that meet the filter

lines 349 to 364

> If we are using a negative filter (i.e. !=) remove those namespaces from the project that meet the filter (i.e. remove stuff from the first step if it matches at least one entry)

365 to 377. We can't simply remove the namespaces from the result, because we are also looking for projects that don't match the endpoint. If I say /namespaces/project1?fieldSelector=prjorns!=project1 then we need to return an empty list.

> Don't do any filtering on project (this is controversial, but I think that if you are accessing this endpoint for a project we should return stuff for the project and only for that project - in this context, filtering by project has no meaning)

See my comments on steve compatibility. This does not return resources for projects that don't match the endpoint.